File: cf-workers/src/push-provider-auth.ts
import { fetchWithTimeout } from "./rocom-client";

const wecomTokenCache = new Map<
  string,
  { token: string; expiresAt: number }
>();

export function clearWecomTokenCache(): void {
  wecomTokenCache.clear();
}

export async function getWecomToken(
  corpid: string,
  secret: string,
  timeoutSec: number
): Promise<string> {
  const key = `${corpid}:${secret}`;
  const cached = wecomTokenCache.get(key);
  const now = Math.floor(Date.now() / 1000);
  if (cached && cached.expiresAt > now + 60) return cached.token;

  const url = `https://qyapi.weixin.qq.com/cgi-bin/gettoken?corpid=${encodeURIComponent(corpid)}&corpsecret=${encodeURIComponent(secret)}`;
  const resp = await fetchWithTimeout(url, { method: "GET" }, timeoutSec);
  const payload = (await resp.json()) as {
    errcode: number | string;
    access_token?: string;
    expires_in?: number;
    errmsg?: string;
  };

  if (resp.status >= 400 || (payload.errcode !== 0 && payload.errcode !== "0")) {
    throw new Error(payload.errmsg || JSON.stringify(payload));
  }

  const token = payload.access_token!;
  const expiresIn = payload.expires_in || 7200;
  wecomTokenCache.set(key, { token, expiresAt: now + expiresIn });
  return token;
}

async function hmacSha256Base64(
  keyData: BufferSource,
  messageData: BufferSource
): Promise<string> {
  const key = await crypto.subtle.importKey(
    "raw",
    keyData,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const sig = await crypto.subtle.sign("HMAC", key, messageData);
  const bytes = new Uint8Array(sig);
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

export async function appendDingTalkSign(
  webhook: string,
  secret: string
): Promise<string> {
  if (!secret) return webhook;
  const timestamp = Date.now().toString();
  const stringToSign = `${timestamp}\n${secret}`;
  const encoder = new TextEncoder();
  const sign = await hmacSha256Base64(
    encoder.encode(secret),
    encoder.encode(stringToSign)
  );
  const sep = webhook.includes("?") ? "&" : "?";
  return `${webhook}${sep}timestamp=${timestamp}&sign=${encodeURIComponent(sign)}`;
}

export async function feishuSign(secret: string, timestamp: string): Promise<string> {
  const stringToSign = `${timestamp}\n${secret}`;
  const encoder = new TextEncoder();
  const key = await crypto.subtle.importKey(
    "raw",
    encoder.encode(stringToSign),
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"]
  );
  const sig = await crypto.subtle.sign("HMAC", key, new Uint8Array(0));
  const bytes = new Uint8Array(sig);
  let binary = "";
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

File: src/roco_push_console/push_provider_auth.py
from __future__ import annotations

import base64
import hashlib
import hmac
import time
import urllib.parse

import requests

from .push_models import ProviderConfig

HttpSession = requests.Session

_WECOM_TOKEN_CACHE: dict[tuple[str, str], tuple[str, float]] = {}


def get_wecom_token(provider: ProviderConfig, session: HttpSession, timeout: int) -> str:
    corpid = provider.config["corpid"]
    secret = provider.config["secret"]
    cache_key = (corpid, secret)
    cached = _WECOM_TOKEN_CACHE.get(cache_key)
    now = time.time()
    if cached and cached[1] > now + 60:
        return cached[0]

    response = session.get(
        "https://qyapi.weixin.qq.com/cgi-bin/gettoken",
        params={"corpid": corpid, "corpsecret": secret},
        timeout=timeout,
    )
    payload = response.json()
    if response.status_code >= 400 or payload.get("errcode") not in (0, "0"):
        raise RuntimeError(str(payload.get("errmsg") or payload))
    token = str(payload["access_token"])
    expires_in = int(payload.get("expires_in", 7200))
    _WECOM_TOKEN_CACHE[cache_key] = (token, now + expires_in)
    return token


def append_dingtalk_sign(webhook: str, secret: str) -> str:
    if not secret:
        return webhook
    timestamp = str(round(time.time() * 1000))
    string_to_sign = f"{timestamp}\n{secret}".encode("utf-8")
    digest = hmac.new(secret.encode("utf-8"), string_to_sign, hashlib.sha256).digest()
    sign = urllib.parse.quote_plus(base64.b64encode(digest))
    separator = "&" if "?" in webhook else "?"
    return f"{webhook}{separator}timestamp={timestamp}&sign={sign}"


def feishu_sign(secret: str, timestamp: str) -> str:
    string_to_sign = f"{timestamp}\n{secret}"
    digest = hmac.new(string_to_sign.encode("utf-8"), b"", digestmod=hashlib.sha256).digest()
    return base64.b64encode(digest).decode("utf-8")
